File: ccollate.rb
#!/usr/bin/env ruby

HEADER_EXTS = %w(.h .hh .hpp)

$include_dirs = ['.']

$visited = []

def parse(file_name)
  if $visited.include? file_name
    return ''
  else
    $visited << file_name
  end

  contents = File.open file_name, 'r' do |file|
    file.read
  end

  contents.gsub!(/\/\*\*(?:.|[\r\n])*?\*\/\s*/, '') # remove documentation comments

  contents = contents.each_line.with_index.collect do |line, lineno|
    lineno = lineno + 1 # line numbers start at 1
    line.gsub!(/^\s*\/\/.*\s*$/, '') # remove c-style comments on their own lines

    if line =~ /^#include\s+"([^"]+)"\s*$/ # recurse and copy
      include_file = nil
      include_dirs = [File.split(file_name).first] + $include_dirs
      include_dirs.each do |inc|
        include_path = File.expand_path File.join(inc, $1)
        if File.exist? include_path
          include_file = include_path
          break
        end
      end

      raise "Fatal Error: The file #{$1} was not found in any of the include paths #{$include_dirs.inspect}" unless include_file

      include = parse(include_file)
      include << "#line #{lineno + 1} \"#{file_name}\"\n"
    else
      line
    end
  end.join
  contents = ["#line 1 \"#{file_name}\"\n", contents].join unless contents.empty?
  contents
end

if ARGV.empty?
  exit(-1)
else
  start_file = File.expand_path ARGV.first
  raise "#{ARGV.first} is not a valid header file." unless HEADER_EXTS.include?(File.extname(start_file))
  $include_dirs << File.split(start_file).first # automatically include the starting directory
  puts parse(start_file)
end

File: ccollate.py
#!/usr/bin/env python

import os
import sys
import re

HEADER_EXTS = ['.h', '.hh', '.hpp']

include_dirs = ['.']

visited = set()

def parse(file_name):
    if file_name in visited:
        return ''
    else:
        visited.add(file_name)

    with open(file_name, 'r') as file:
        contents = file.read()

    # Remove documentation comments
    contents = re.sub(r'/\*\*(?:.|[\r\n])*?\*/\s*', '', contents)

    contents = '\n'.join([line for line in contents.splitlines() if not line.strip().startswith('//')])

    lines = contents.splitlines()
    contents = []

    for lineno, line in enumerate(lines, start=1):
        if line.startswith('#include "'):
            include_file = None
            include_dirs.insert(0, os.path.dirname(file_name))
            for inc in include_dirs:
                include_path = os.path.abspath(os.path.join(inc, line.split('"')[1]))
                if os.path.exists(include_path):
                    include_file = include_path
                    break

            if include_file is None:
                name = line.split('"')[1]
                raise Exception(f"Fatal Error: The file {name} was not found in any of the include paths {include_dirs}")

            include = parse(include_file)
            include = f"#line {lineno + 1} \"{file_name}\"\n{include}"
            contents.append(include)
        else:
            contents.append(line)

    contents = '\n'.join(contents)
    if contents:
        contents = f"#line 1 \"{file_name}\"\n{contents}"
    return contents

if len(sys.argv) == 1:
    sys.exit(-1)
else:
    start_file = os.path.abspath(sys.argv[1])
    if not os.path.isfile(start_file) or os.path.splitext(start_file)[1] not in HEADER_EXTS:
        raise Exception(f"{sys.argv[1]} is not a valid header file.")
    include_dirs.append(os.path.dirname(start_file))
    print(parse(start_file))
